data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java

```diff
@@ -196,24 +199,14 @@ public GetForkResponse getForkInfo() {
       return chainUnavailable();
     }
 
-    final Optional<UInt64> maybeSlot = blockParameterToSlot(slotParameter);
-    if (maybeSlot.isEmpty()) {
-      return getBlockHeaderByBlockRoot(Bytes32.fromHexString(slotParameter));
-    }
-
-    final UInt64 slot = maybeSlot.get();
-    return combinedChainDataClient
-        .getBlockAtSlotExact(slot)
-        .thenApply(maybeBlock -> maybeBlock.map(block -> new BlockHeader(block, true)));
-  }
-
-  // because this is called after attempting to match a block to a slot, this function
-  // will only ever be run for non canonical blocks. if made public, it will have to be updated to
-  // first check that the block doesnt have a slot, before it can make that assumption.
-  SafeFuture<Optional<BlockHeader>> getBlockHeaderByBlockRoot(final Bytes32 blockRoot) {
-    return combinedChainDataClient
-        .getBlockByBlockRoot(blockRoot)
-        .thenApply(maybeBlock -> maybeBlock.map(block -> new BlockHeader(block, false)));
+    return defaultBlockSelectorFactory
+        .defaultBlockSelector(slotParameter)
+        .getBlock()
+        .thenApply(
+            blockList ->
+                blockList.isEmpty()
+                    ? Optional.empty()
+                    : Optional.of(new BlockHeader(blockList.get(0), true)));
```

Comment:
We could have a default method `getSingleBlock` in the `BlockSelector` interface that does this conversion from list to `Optional`.  Could just do a `blockList.stream().findFirst()` so simplify it.

Reply:
Also I'd have expected the handler to convert the String slotParameter to a BlockSelector and then pass the BlockSelector into this method.  That way the handler is responsible for interpreting the request and the provider has the business logic.  With it this way the provider is interpreting the request and doing the business logic.